### PR TITLE
[Consensus] State sync design overhaul - block-by-block state sync

### LINKED ISSUE #362: [Consensus] State Sync - Define Block By Block Interface
## Objective

Update the interface in `consensus/state_sync/state_sync.go` so it is well-defined enough for the implementation to be picked up by anyone.

## Origin Document

This is a follow-up to #125 per the comments & discussion in #342.

The list of changes this encompasses includes (but is not limited to):
- Remove callback functions and use the application-specific bus for everything
- Adding protobuf definitions
- Splitting the different `Run` functions into separate modules
- Updating terminology if applicable

## Goals

- Finalize the block-by-block state sync design
- Make the interface clear enough such that it is implementable by anyone

## Deliverable

- [ ] Updated interface internal to the consensus module
- [ ] Updated interface in the shared directory
- [ ] Updated protobufs where applicable

## Non-goals / Non-deliverables

- Implement State Sync

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

## Testing Methodology

- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @Olshansk

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

Defined the interface for block-by-block state sync interface along with documentation updates.

## Issue

Fixes #362

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [X] Documentation
- [ ] Other 

## List of changes

- Added protobuf message definitions for requests related to sharing state sync metadata and blocks
- Defined the interface for `StateSyncServerModule`, `StateSyncModule` (moving the old interface to `StateSyncModuleLEGACY` as a reference only)
- Overhaul (updates, improvements, clarifications & additions) of the State Sync README

## Testing

- [x] `make develop_test`
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by Olshansk** [comment]
@gokutheengineer Please note that there will be implementation details in #409 that are not covered and discussed here (e.g. the local state consensus will be managing as blocks are flying in and we're processing them), so lmk if this requires any additional detail.

**Review by gokutheengineer — CHANGES_REQUESTED** [review]

File: consensus/doc/PROTOCOL_STATE_SYNC.md [hunk 1]
```diff
@@ -1,119 +1,409 @@
...
+        NP->>N: Yup, here's my Peer Metadata. What's yours?
+        N->>NP: ACK, here's mine. I'll ask again in a bit to make sure I'm up to date.
     end
 ```
 
-The aggregation and consumption of this peer-meta information enables the State Sync protocol by enabling the node to understand the globalized network state by sampling Peer Metadata through its local peer list.
-This gives a view into the data availability layer, with details of what data can be consumed from which peer.
+The aggregation and consumption of this peer-meta information enables the State Sync protocol by enabling the node to understand the globalized network state through sampling Peer Metadata in its local peer list.
+
+This gives a view into the data availability layer, with details of what data can be consumed from which pee (not a production interface):
```
Comment (Olshansk) [inline consensus/doc/PROTOCOL_STATE_SYNC.md:76]:
```suggestion
This gives a view into the data availability layer, with details of what data can be consumed from which peer (not a production interface):
```

[the remark this answers is not in the document]
Comment (Olshansk) [reply consensus/doc/PROTOCOL_STATE_SYNC.md]: Resolved here: https://github.com/pokt-network/pocket/pull/413#discussion_r1059107713

Also, highly recommend using the "suggestion" feature available here for edits in the future:
Screenshot 2022-12-29 at 1 59 32 PM

File: consensus/doc/PROTOCOL_STATE_SYNC.md [hunk 2]
```diff
@@ -1,119 +1,409 @@
...
+    loop Churn Management
+        N->>+NP: Are you alive? If so, what's your Peer Metadata?
+        NP->>N: Yup, here's my Peer Metadata. What's yours?
+        N->>NP: ACK, here's mine. I'll ask again in a bit to make sure I'm up to date.
     end
 ```
 
-The aggregation and consumption of this peer-meta information enables the State Sync protocol by enabling the node to understand the globalized network state by sampling Peer Metadata through its local peer list.
-This gives a view into the data availability layer, with details of what data can be consumed from which peer.
+The aggregation and consumption of this peer-meta information enables the State Sync protocol by enabling the node to understand the globalized network state through sampling Peer Metadata in its local peer list.
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply consensus/doc/PROTOCOL_STATE_SYNC.md:74]: Is the sequenceDiagram above not sufficient? At a high level, it's: "Get peer metadata from peers," and the implementation details involve aggregating it, creating triggers, etc...

I explicitly chose to exclude it from this PR and was hoping that it's something you'd be able to add to the documentation in a couple of iterations once we get closer to designing that (after implementing the basic advertisement and retrieval) of peer metadata.

Wdyt?

[the remark this answers is not in the document]
Comment (gokutheengineer) [reply consensus/doc/PROTOCOL_STATE_SYNC.md:74]: Okay, I will add once we have better idea.

**Review by gokutheengineer — APPROVED** [review]

File: consensus/doc/PROTOCOL_STATE_SYNC.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply consensus/doc/PROTOCOL_STATE_SYNC.md:74]: We can do a group design session in the future. I think that once we have the implementation for advertising & retrieving (plus the PR that Alessandro has for peer discovery), it'll be clearer.

File: Makefile [hunk 3]
```diff
@@ -238,6 +238,7 @@ protogen_clean: ## Remove all the generated protobufs.
 
 .PHONY: protogen_local
 protogen_local: go_protoc-go-inject-tag ## Generate go structures for all of the protobufs
+# TODO: Organize this code with a basic fo
```
Comment (deblasis) [inline Makefile:241]:
```suggestion
# TODO: Organize this code with a basic for loop
```
?

P.S. I believe that @Gustavobelfort was looking into integrating `buf` and at that point protobuf generation would be more declarative and without all these calls. Just flagging

File: consensus/doc/PROTOCOL_STATE_SYNC.md [hunk 4]
```diff
@@ -1,119 +1,409 @@
...
+
+### Future Design Work
+
+- `Fast Sync Design` - Sync only the last `N` blocks from a _snapshot_ containing a network state
 
-How the persistence layer design of `pruning` the Merkle Tree affects `StateSync`.
+- `Optimistic Sync Design` - Optimize the State Sync protocol by parallelling requests and computation with pre-fetching and local caching
 
-How DB Pruning of the `SQL DB` might affect fast sync.
+- `Block Chunk Design` - Update the singe block-by-block to be able to receive and provide multiple blocks per request.
```
Comment (deblasis) [inline consensus/doc/PROTOCOL_STATE_SYNC.md:211]:
```suggestion
- `Block Chunk Design` - Update the single block-by-block to be able to receive and provide multiple blocks per request.
```

**Review by deblasis — APPROVED** [review]
LGTM, just nits

**Pull request merged**